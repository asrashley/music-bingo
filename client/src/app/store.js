import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import { connectRouter } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import { reducer as formReducer } from 'redux-form';

import cardsReducer from '../cards/cardsSlice';
import gamesReducer from '../games/gamesSlice';
import ticketsReducer from '../tickets/ticketsSlice';
import userReducer from '../user/userSlice';
import adminReducer from '../admin/adminSlice';
import usersMiddleware from '../user/userMiddleware';

export const history = createBrowserHistory();

export const store = configureStore({
  middleware: [...getDefaultMiddleware(), usersMiddleware, logger],
  reducer: {
    admin: adminReducer,
    cards: cardsReducer,
    form: formReducer,
    games: gamesReducer,
    tickets: ticketsReducer,
    router: connectRouter(history),
    user: userReducer,
  },
});
