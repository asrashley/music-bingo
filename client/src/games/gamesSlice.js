import { createSlice } from '@reduxjs/toolkit';

import { api } from '../endpoints';

const gameAdditionalFields = {
  tracks: [],
  isFetchingDetail: false,
  invalidDetail: true,
  lastUpdated: null,
  isModifying: false,
};
Object.freeze(gameAdditionalFields);

const dateOrder = (a, b) => {
  if (a.start < b.start) {
    return -1;
  } else if (a.start > b.start) {
    return 1;
  }
  return 0;
};

export const gamesSlice = createSlice({
  name: 'games',
  initialState: {
    games: {},
    order: [],
    pastOrder: [],
    user: -1,
    isFetching: false,
    invalid: true,
    error: null,
    lastUpdated: null,
  },
  reducers: {
    invalidateGames: state => {
      state.invalid = true;
    },
    requestGames: state => {
      state.isFetching = true;
    },
    receiveGames: (state, action) => {
      const { timestamp, payload, user } = action.payload;
      const { games, past } = payload;
      state.isFetching = false;
      state.error = null;
      state.lastUpdated = timestamp;
      state.invalid = false;
      state.user = user.pk;
      state.games = {};
      state.order = [];
      state.pastOrder = [];
      games.forEach(game => {
        state.games[game.pk] = {
          ...game,
          ...gameAdditionalFields,
        };
        state.order.push(game.pk);
      });
      past.forEach(game => {
        state.games[game.pk] = {
          ...game,
          ...gameAdditionalFields,
        };
        state.pastOrder.push(game.pk);
      });
    },
    failedFetchGames: (state, action) => {
      const { timestamp, error } = action.payload;
      state.isFetching = false;
      state.lastUpdated = timestamp;
      state.error = error;
      state.invalid = true;
    },
    requestDetail: (state, action) => {
      const { gamePk } = action.payload;
      if (!state.games[gamePk]) {
        return;
      }
      state.games[gamePk].isFetchingDetail = true;
    },
    modifyGame: (state, action) => {
      const { gamePk } = action.payload;
      if (!state.games[gamePk]) {
        return;
      }
      state.games[gamePk].isModifying = true;
    },
    failedModifyGame: (state, action) => {
      const { gamePk } = action.payload;
      if (!state.games[gamePk]) {
        return;
      }
      state.games[gamePk].isModifying = false;
    },
    receiveDetail: (state, action) => {
      const { timestamp, payload } = action.payload;
      if (!state.games[payload.pk]) {
        console.log('failed find game PK');
        console.dir(payload);
        return;
      }
      state.games[payload.pk] = {
        ...state.games[payload.pk],
        ...payload,
        isFetchingDetail: false,
        invalidDetail: false,
        lastUpdated: timestamp,
      };
    },
    receiveGameModification: (state, action) => {
      const { timestamp, payload, gamePk } = action.payload;
      if (!state.games[gamePk]) {
        return;
      }
      const updateOrder = state.games[gamePk].start !== payload.game.start ||
        state.games[gamePk].end !== payload.game.end;
      state.games[gamePk] = {
        ...state.games[gamePk],
        ...payload.game,
        isModifying: false,
        lastUpdated: timestamp,
      };
      if (updateOrder) {
        const order = [];
        const pastOrder = [];
        const now = new Date().toISOString();
        let today = new Date();
        today.setHours(0);
        today.setMinutes(0);
        today = today.toISOString();
        for (let pk in state.games) {
          const game = state.games[pk];
          if (game.start >= today && game.end > now) {
            order.push(game);
          } else {
            pastOrder.push(game);
          }
        }
        order.sort(dateOrder);
        pastOrder.sort(dateOrder);
        state.order = order.map(g => g.pk);
        state.pastOrder = pastOrder.map(g => g.pk);
      }
    },
    receiveGameDeleted: (state, action) => {
      const { gamePk } = action.payload;
      if (!state.games[gamePk]) {
        return;
      }
      delete state.games[gamePk];
      state.order = state.order.filter(pk => pk !== gamePk);
      state.pastOrder = state.pastOrder.filter(pk => pk !== gamePk);
    },
    failedFetchDetail: (state, action) => {
      const { gamePk, error, timestamp } = action.payload;
      if (!state.games[gamePk]) {
        return;
      }
      state.games[gamePk] = {
        ...state.games[gamePk],
        isFetchingDetail: false,
        invalidDetail: true,
        lastUpdated: timestamp,
        error,
      };
    },
    invalidateGameDetail: (state, action) => {
      const { gamePk } = action.payload;
      if (!state.games[gamePk]) {
        return;
      }
      state.games[gamePk] = {
        ...state.games[gamePk],
        tracks: [],
        isFetchingDetail: false,
        invalidDetail: true,
      };
    },

  },
});

function fetchGames(userId) {
  return api.getGamesList({
    before: gamesSlice.actions.requestGames,
    failure: gamesSlice.actions.failedFetchGames,
    success: gamesSlice.actions.receiveGames,
  });
}

function shouldFetchGames(state) {
  const { games, user } = state;
  if (games.isFetching) {
    return false;
  }
  return games.invalid || user.pk !== games.user;
}

export function fetchGamesIfNeeded() {
  return (dispatch, getState) => {
    const state = getState();
    if (shouldFetchGames(state)) {
      return dispatch(fetchGames(state.user.pk));
    }
    return Promise.resolve();
  };
}

function fetchDetail(userPk, gamePk) {
  return api.getGameDetail({
    gamePk,
    before: gamesSlice.actions.requestDetail,
    failure: gamesSlice.actions.failedFetchDetail,
    success: gamesSlice.actions.receiveDetail
  });
}

function shouldFetchDetail(state, gamePk) {
  const { games, user } = state;
  const game = games.games[gamePk];
  if (!game) {
    return false;
  }
  if (user.pk !== games.user) {
    return true;
  }
  if (game.isFetchingDetail) {
    return false;
  }
  return game.invalidDetail;
}

export function fetchDetailIfNeeded(gamePk) {
  return (dispatch, getState) => {
    const state = getState();
    if (shouldFetchDetail(state, gamePk)) {
      return dispatch(fetchDetail(state.user.pk, gamePk));
    }
    const game = state.games.games[gamePk];
    return Promise.resolve(game ? game : null);
  };
}

export function modifyGame(game) {
  const gamePk = game.pk;
  return api.modifyGame({
    gamePk,
    body: game,
    before: gamesSlice.actions.modifyGame,
    success: gamesSlice.actions.receiveGameModification,
    failure: gamesSlice.actions.failedModifyGame,
  });
}

export function deleteGame(game) {
  const gamePk = game.pk;
  return api.deleteGame({
    body: game,
    gamePk,
    before: gamesSlice.actions.modifyGame,
    success: gamesSlice.actions.gameDeleted,
    failure: gamesSlice.actions.failedModifyGame,
  });
}

export const { invalidateGames, invalidateGameDetail } = gamesSlice.actions;

export const initialState = gamesSlice.initialState;

export default gamesSlice.reducer;
