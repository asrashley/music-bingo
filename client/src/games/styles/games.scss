@import "../../styles/brand";
@use "sass:list";

#games-page {
  .greeting {
    margin: 1rem 0 2rem 1rem;
  }
}

.game-list {
  position: relative;

  td.round-column {
    text-align: center;
  }

  .round-1 > td {
    padding-top: 1em;
  }

  .round-column {
    width: 5em;
  }

  .date-column {
    width: 7em;
  }

  .theme-column {
    width: 28em;
  }


  .ticket-column {
    width: 15em;
  }
}

.track-listing {
  padding-bottom: 1em;

  .number {
    text-align: center;
    width: 4em;
  }

  .start-time,
  .duration {
    width: 5em;
  }

  td.start-time,
  td.duration {
    text-align: right;
  }

  .title,
  .artist,
  .album {
    width: 24em;
  }

  .title {
    font-size: 90%;
  }

  .artist,
  .album {
    font-size: 80%;
    font-stretch: condensed;
  }
}

.modify-game {
  margin: 1em 0 2em 0;
  max-width: 50em;
  padding: 1em 2em;
  position: relative;
}

.game-list .refresh-icon,
.modify-game .refresh-icon {
  position: absolute;
  right: 0;
  top: 0;
}

.horiz-popularity-graph {
  margin-bottom: 2em;

  .title {
    padding-bottom: 0.25em;
    padding-right: 1em;
    text-align: right;
    width: 15em;
  }

  tbody .count {
    color: $bar-chart-number;
    padding-bottom: 0.25em;
    text-align: center;
    width: 30em;
  }

  .rotate-icon {
    background-image: url("../img/vert-chart.png");
    background-size: cover;
    float: right;
    height: 2em;
    width: 3em;
  }
}

.vert-popularity-graph {
  display: none;
  font-size: 0.75rem;
  width: 100%;

  .bar-wrap {
    height: 20em;
    position: relative;
  }

  .theme {
    display: inline-block;
    margin: 0.1em;
  }

  .bar {
    bottom: 0;
    color: white;
    padding-bottom: 0.5em;
    position: absolute;
    text-align: right;
    text-orientation: sideways;
    writing-mode: vertical-rl;
  }

  .name {
    height: 15em;
    margin: 0.1em;
    overflow: hidden;
    padding-top: 0.5rem;
    text-orientation: sideways;
    writing-mode: vertical-rl;
  }

  .bar,
  .name,
  .theme {
    line-height: 1em;
    width: 1em;
  }

  .rotate-icon {
    background-image: url("../img/horiz-chart.png");
    background-size: cover;
    float: right;
    height: 3em;
    width: 2em;
  }
}

.file-dialog {
  .form-control {
    margin: 1em 0.25em;
    padding: 0.25em 0.75em;
  }

  .replace-date-label {
    margin-right: 1em;
  }
}

@media (min-width: 700px) {
  .vert-popularity-graph {
    display: block;
  }
}

@media (min-width: 700px) {
  .vert-popularity-graph {
    font-size: 0.8rem;
  }
}

@media (min-width: 980px) {
  .vert-popularity-graph {
    font-size: 0.85rem;
  }

  .vert-popularity-graph .name,
  .vert-popularity-graph .theme,
  .vert-popularity-graph .bar {
    line-height: 1.25em;
    width: 1.25em;
  }
}

@media (min-width: 1200px) {
  .vert-popularity-graph .name,
  .vert-popularity-graph .theme,
  .vert-popularity-graph .bar {
    line-height: 1.5em;
    width: 1.5em;
  }
}
