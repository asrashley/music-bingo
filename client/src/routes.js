
export default {
  index: '/',
  login: '/user/login',
  logout: '/user/logout',
  passwordReset: '/user/reset',
  passwordResetConfirm: '/user/reset/:token',
  register: '/register',
  game: '/game/:gamePk',
  play: '/game/:gamePk/tickets',
  viewTicket: '/game/:gamePk/:ticketPk',
  pastGames: '/history',
  trackListing: '/history/:gamePk',
  listUsers: '/users'
};

