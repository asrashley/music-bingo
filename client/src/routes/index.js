export default {
  index: '/',
  login: '/user/login',
  logout: '/user/logout',
  user: '/user',
  changeUser: '/user/modify',
  passwordReset: '/user/reset',
  passwordResetConfirm: '/user/reset/:token',
  guestLinks: '/user/guests',
  listUsers: '/user/users',
  guestAccess: '/invite/:token',
  register: '/register',
  listDirectories: '/directories',
  listGames: '/game',
  chooseTickets: '/game/:gameId',
  play: '/game/:gameId/tickets',
  viewTicket: '/game/:gameId/:ticketPk',
  pastGames: '/history',
  trackListing: '/history/:gameId',
  privacy: '/privacy',
};
