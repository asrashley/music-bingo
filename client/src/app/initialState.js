import { initialState as cards } from '../cards/cardsSlice';
import { initialState as games } from '../games/gamesSlice';
import { initialState as tickets } from '../tickets/ticketsSlice';
import { initialState as user } from '../user/userSlice';

export const initialState = Object.freeze({
    cards,
    games,
    tickets,
    user,
});
