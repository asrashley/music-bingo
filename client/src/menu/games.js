import { reverse } from 'named-urls';

import routes from '../routes';

const gamesMenuItems = [{
    href: reverse(`${routes.pastGamesPopularity}`),
    title: 'Popularity',
}, {
    href: reverse(`${routes.gameLastUsed}`),
    title: "Theme's Last Use",
}, {
    href: reverse(`${routes.pastGamesCalendar}`),
    title: 'Calendar',
}, {
    href: reverse(`${routes.pastGamesList}`),
    title: 'All Previous Games',
}];

export function getGameMenuItems(/* state, props */) {
    return gamesMenuItems;
}
