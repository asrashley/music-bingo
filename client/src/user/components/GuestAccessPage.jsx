import React, { useEffect, useMemo } from 'react';
import PropTypes from 'prop-types';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useParams } from 'react-router-dom';
import { reverse } from 'named-urls';
import { push } from '@lagunovsky/redux-react-router';

import { Welcome } from '../../app/Welcome';
import {
  fetchUserIfNeeded, checkGuestTokenIfNeeded, createGuestAccount, loginUser,
  clearGuestDetails,
} from '../../user/userSlice';
import { getUser } from '../../user/userSelectors';
import { routes } from '../../routes/routes';
import { gameInitialFields } from '../../games/gamesSlice';
import { ticketInitialState } from '../../tickets/ticketsSlice';
import { GuestPropType } from '../types/Guest';

import '../styles/user.scss';

function GuestStatusMessage({ token, guest }) {
  if (token === undefined || guest?.valid === false) {
    return (
      <React.Fragment>
        <p className="description alert alert-warning">Sorry, the link you have used is not recognized</p>
        <p className="description">Either double-check the guest link you received, login or create
          an account.</p>
      </React.Fragment>
    );
  }
  if (guest?.isFetching === true || guest?.valid === null) {
    return (
      <h2 className="description">Checking if guest link is valid...</h2>
    );
  }
  if (guest?.error) {
    return (
      <div className="alert alert-warning" role="alert">
        <span className="error-message">{guest.error}</span>
      </div>
    );
  }
  return null;
}

GuestStatusMessage.propTypes = {
  token: PropTypes.string,
  guest: GuestPropType,
};

export function GuestAccessPage({ nextPage = reverse(`${routes.index}`) }) {
  const [game, ticket] = useMemo(() => {
    const now = new Date();
    const fakeGame = {
      ...gameInitialFields,
      id: `${now.getFullYear()}-${now.getMonth()}-${now.getDate()}`,
    };
    const fakeTicket = {
      ...ticketInitialState(),
      number: 1 + now.getHours(),
    };
    return [fakeGame, fakeTicket];
  }, []);
  const { token } = useParams();
  const dispatch = useDispatch();
  const user = useSelector(getUser);
  const { guest } = user;

  useEffect(() => {
    dispatch(fetchUserIfNeeded());
  }, [dispatch]);

  useEffect(() => {
    if (user.loggedIn === true) {
      dispatch(push(nextPage));
      return;
    }
  }, [dispatch, user.loggedIn, nextPage]);

  useEffect(() => {
    // TODO: check if token needs validation
    if (token) {
      dispatch(checkGuestTokenIfNeeded(token));
    }
  }, [dispatch, token]);

  const playAsGuest = async () => {
    if (!guest?.username || !guest?.password) {
      dispatch(createGuestAccount(token, nextPage));
      return;
    }
    const { payload } = await dispatch(loginUser({
      password: guest.password,
      username: guest.username,
      rememberme: false,
    }));
    if (payload && payload.accessToken) {
      dispatch(push(nextPage));
    } else {
      dispatch(clearGuestDetails());
    }
  };

  return (
    <Welcome className="guest-link index-page" game={game} ticket={ticket}>
      <GuestStatusMessage token={token} guest={guest} />
      <p className="action-buttons mt-5">
        <Link to={reverse(`${routes.login}`)} className="login nav-link btn btn-success btn-lg login-button">Log in</Link>
        <Link to={reverse(`${routes.register}`)} className="login nav-link btn btn-success btn-lg login-button">Create an account</Link>
        {guest.valid === true && <button className="play-guest nav-link btn btn-primary btn-lg"
          onClick={playAsGuest}>Play as a guest</button>}
      </p>
    </Welcome>
  );
}

GuestAccessPage.propTypes = {
  nextPage: PropTypes.string,
};
