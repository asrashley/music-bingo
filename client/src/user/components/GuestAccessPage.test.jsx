import React from 'react';
import { screen } from '@testing-library/react';
import fetchMock from "fetch-mock-jest";
import log from 'loglevel';
import { reverse } from 'named-urls';

import { renderWithProviders, installFetchMocks } from '../../testHelpers';
import { GuestAccessPage, GuestAccessPageComponent } from './GuestAccessPage';
import routes from '../../routes';
import { initialState } from '../../store/initialState';
import { gameInitialFields } from '../../games/gamesSlice';
import { ticketInitialState } from '../../tickets/ticketsSlice';

describe('GuestAccessPage component', () => {
  let apiMocks;

  beforeAll(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('08 Feb 2023 10:12:00 GMT').getTime());
  });

  afterEach(() => {
    fetchMock.mockReset();
    log.resetLevel();
    apiMocks = null;
  });

  afterAll(() => jest.useRealTimers());

  beforeEach(() => {
    apiMocks = installFetchMocks(fetchMock, { loggedIn: false });
  });

  it('allows playing as a guest if token is valid', async () => {
    const user = {
      ...initialState.user,
      guest: {
        valid: true,
        isFetching: false
      }
    };
    const preloadedState = {
      ...initialState,
      user,
    };
    const match = {
      params: {
        token: '123abc'
      }
    };
    const history = {
      push: jest.fn()
    };
    fetchMock.post('/api/user/guest', (url, opts) => {
      return apiMocks.jsonResponse({
        ...user,
        success: true,
        username: 'guest005',
        accessToken: 'guest.access.token',
        password: 'random.secret',
        expires: 3600,
        refreshToken: 'guest.refresh.token',
      });
    });
    const { asFragment } = renderWithProviders(<GuestAccessPage match={match} history={history} />, { preloadedState });
    await screen.findByText('Play as a guest');
    expect(asFragment()).toMatchSnapshot();
  });

  const renderGuestAccessPageComponent = (guest) => {
    const history = {
      push: jest.fn()
    };
    const game = {
      ...gameInitialFields,
      id: '2023-03-10'
    };
    const ticket = {
      ...ticketInitialState(),
      number: 21,
    };
    const props = {
      dispatch: jest.fn(),
      game,
      token: '123abc',
      history,
      ticket,
      user: {
        ...initialState.user,
        guest
      }
    };
    return renderWithProviders(<GuestAccessPageComponent {...props} />);
  };

  it('shows a message if token is being checked', () => {
    const guest = {
      valid: true,
      isFetching: true
    };
    const result = renderGuestAccessPageComponent(guest);
    result.getByText('Checking if guest link is valid...');
  });

  it('shows an error if token is invalid', () => {
    const guest = {
      valid: false,
      isFetching: false
    };
    const result = renderGuestAccessPageComponent(guest);
    result.getByText('Sorry, the link you have used is not recognised');
  });

  it('shows a generic error', () => {
    const guest = {
      valid: true,
      isFetching: false,
      error: 'a generic error'
    };
    const result = renderGuestAccessPageComponent(guest);
    result.getByText(guest.error);
  });

  it('redirects if user is logged in', () => {
    const history = {
      push: jest.fn()
    };
    const game = {
      ...gameInitialFields,
      id: '2023-03-10'
    };
    const ticket = {
      ...ticketInitialState(),
      number: 21,
    };
    const props = {
      dispatch: jest.fn(),
      game,
      token: '123abc',
      history,
      ticket,
      user: {
        ...initialState.user,
        loggedIn: true
      }
    };
    renderWithProviders(<GuestAccessPageComponent {...props} />);
    expect(history.push).toHaveBeenCalledTimes(1);
  });

  it('calls login if guest username and password are provided', async () => {
    const user = {
      ...initialState.user,
      guest: {
        valid: true,
        isFetching: false,
        username: 'guest123',
        email: 'guest123',
        password: 'random.secret'
      }
    };
    const preloadedState = {
      ...initialState,
      user,
    };
    const match = {
      params: {
        token: '123abc'
      }
    };
    fetchMock.post('/api/user/guest', (url, opts) => {
      return apiMocks.jsonResponse({
        ...user,
        success: true,
        username: user.guest.username,
        accessToken: apiMocks.getAccessToken(),
        password: user.guest.password,
        expires: 3600,
        refreshToken: 'guest.refresh.token',
      });
    });
    apiMocks.addUser(user.guest);
    const url = await new Promise((resolve) => {
      const history = {
        push: resolve
      };
      renderWithProviders(<GuestAccessPage match={match} history={history} />, { preloadedState });
    });
    expect(url).toEqual(reverse(`${routes.index}`));
  });

});
