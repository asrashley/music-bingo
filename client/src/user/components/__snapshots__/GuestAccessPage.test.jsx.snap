// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GuestAccessPage component > allows playing as a guest if token is valid 1`] = `
<DocumentFragment>
  <div
    class="guest-link index-page"
  >
    <div
      class="logo"
    />
    <div
      class="welcome"
    >
      <h2
        class="strapline"
      >
        Like normal Bingo, but with music!
      </h2>
      <p
        class="description"
      >
        Musical Bingo is a variation on the normal game of bingo, where the numbers are replaced with songs that the players must listen out for.
      </p>
      <p
        class="description"
      >
        If you know your Bruce Springsteen from your Beyonce, your Whitney Houston from your Wu-Tang Clan, why not join use for a game or two?
      </p>
      <h2
        class="description"
      >
        Checking if guest link is valid...
      </h2>
      <p
        class="action-buttons mt-5"
      >
        <a
          class="login nav-link btn btn-success btn-lg login-button"
          href="/user/login"
        >
          Log in
        </a>
        <a
          class="login nav-link btn btn-success btn-lg login-button"
          href="/register"
        >
          Create an account
        </a>
        <button
          class="play-guest nav-link btn btn-primary btn-lg"
        >
          Play as a guest
        </button>
      </p>
    </div>
    <div
      class="number footer"
    >
      Game 2023-1-8 / Ticket 11
    </div>
  </div>
</DocumentFragment>
`;

exports[`GuestAccessPage component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="guest-link index-page"
  >
    <div
      class="logo"
    />
    <div
      class="welcome"
    >
      <h2
        class="strapline"
      >
        Like normal Bingo, but with music!
      </h2>
      <p
        class="description"
      >
        Musical Bingo is a variation on the normal game of bingo, where the numbers are replaced with songs that the players must listen out for.
      </p>
      <p
        class="description"
      >
        If you know your Bruce Springsteen from your Beyonce, your Whitney Houston from your Wu-Tang Clan, why not join use for a game or two?
      </p>
      <h2
        class="description"
      >
        Checking if guest link is valid...
      </h2>
      <p
        class="action-buttons mt-5"
      >
        <a
          class="login nav-link btn btn-success btn-lg login-button"
          href="/user/login"
        >
          Log in
        </a>
        <a
          class="login nav-link btn btn-success btn-lg login-button"
          href="/register"
        >
          Create an account
        </a>
        <button
          class="play-guest nav-link btn btn-primary btn-lg"
        >
          Play as a guest
        </button>
      </p>
    </div>
    <div
      class="number footer"
    >
      Game 2023-1-8 / Ticket 11
    </div>
  </div>
</DocumentFragment>
`;

exports[`GuestAccessPage component allows playing as a guest if token is valid 1`] = `
<DocumentFragment>
  <div
    class="guest-link index-page"
  >
    <div
      class="logo"
    />
    <div
      class="welcome"
    >
      <h2
        class="strapline"
      >
        Like normal Bingo, but with music!
      </h2>
      <p
        class="description"
      >
        Musical Bingo is a variation on the normal game of bingo, where the numbers are replaced with songs that the players must listen out for.
      </p>
      <p
        class="description"
      >
        If you know your Bruce Springsteen from your Beyonce, your Whitney Houston from your Wu-Tang Clan, why not join use for a game or two?
      </p>
      <p
        class="action-buttons mt-5"
      >
        <a
          class="login nav-link btn btn-success btn-lg login-button"
          href="/user/login"
        >
          Log in
        </a>
        <a
          class="login nav-link btn btn-success btn-lg login-button"
          href="/register"
        >
          Create an account
        </a>
        <button
          class="play-guest nav-link btn btn-primary btn-lg"
        >
          Play as a guest
        </button>
      </p>
    </div>
    <div
      class="number footer"
    >
      Game 2023-1-8 / Ticket 11
    </div>
  </div>
</DocumentFragment>
`;
